{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "98002a45-2424-44eb-bc1c-3b1fb3f33668",
   "metadata": {},
   "source": [
    "Q1. Explain why we have to use the Exception class while creating a Custom Exception.                                     \n",
    "Note: Here Exception class refers to the base class for all the exceptions."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "256bc838-8dc4-4b13-b5cc-80d150e24f0a",
   "metadata": {},
   "source": [
    "#### When creating a custom exception in a programming language, it is common practice to derive it from the base exception class provided by the language, often called the \"Exception\" class. The Exception class serves as the foundation for all exceptions in the language, and using it as the base for custom exceptions offers several benefits:\n",
    "\n",
    "#### 1) Standardization: By using the Exception class as the base for custom exceptions, we adhere to a standardized exception hierarchy within the language. This consistency makes it easier for developers to understand and handle exceptions throughout different parts of the codebase. It also promotes code readability and maintainability, as other developers will be familiar with the exception hierarchy and its behavior.\n",
    "\n",
    "#### 2) Exception Handling: The Exception class provides a rich set of properties and methods that facilitate exception handling. These include properties such as the exception message, stack trace, and inner exception, as well as methods like ToString(), GetType(), and GetBaseException(). By deriving from the Exception class, our custom exception inherits these capabilities, making it easier to capture and report relevant information when an exception occurs. It also allows developers to handle our custom exception using the same exception handling mechanisms used for built-in exceptions.\n",
    "\n",
    "#### 3) Compatibility: Many libraries, frameworks, and third-party tools are designed to work with exceptions derived from the Exception class. By deriving our custom exception from this base class, we increase the likelihood that our exception will be compatible with existing exception handling mechanisms, error logging frameworks, and other tools. This compatibility can save development time and effort, as we can leverage existing infrastructure for exception handling and error reporting.\n",
    "\n",
    "#### 4) Documentation and Community: The Exception class is usually well-documented and widely used within the programming language's community. By using it as the base class for our custom exception, we can benefit from the existing documentation, examples, and best practices associated with handling exceptions in the language. This can be particularly helpful when troubleshooting issues or seeking assistance from the community."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e79b282-e1a0-4430-b6e5-c12521a8d318",
   "metadata": {},
   "source": [
    "Q2. Write a python program to print Python Exception Hierarchy."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "44be897c-2f93-43c4-9809-b57e40562b37",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BaseException\n",
      "    Exception\n",
      "        TypeError\n",
      "            FloatOperation\n",
      "            MultipartConversionError\n",
      "        StopAsyncIteration\n",
      "        StopIteration\n",
      "        ImportError\n",
      "            ModuleNotFoundError\n",
      "            ZipImportError\n",
      "        OSError\n",
      "            ConnectionError\n",
      "                BrokenPipeError\n",
      "                ConnectionAbortedError\n",
      "                ConnectionRefusedError\n",
      "                ConnectionResetError\n",
      "                    RemoteDisconnected\n",
      "            BlockingIOError\n",
      "            ChildProcessError\n",
      "            FileExistsError\n",
      "            FileNotFoundError\n",
      "            IsADirectoryError\n",
      "            NotADirectoryError\n",
      "            InterruptedError\n",
      "                InterruptedSystemCall\n",
      "            PermissionError\n",
      "            ProcessLookupError\n",
      "            TimeoutError\n",
      "            UnsupportedOperation\n",
      "            itimer_error\n",
      "            herror\n",
      "            gaierror\n",
      "            SSLError\n",
      "                SSLCertVerificationError\n",
      "                SSLZeroReturnError\n",
      "                SSLWantWriteError\n",
      "                SSLWantReadError\n",
      "                SSLSyscallError\n",
      "                SSLEOFError\n",
      "            Error\n",
      "                SameFileError\n",
      "            SpecialFileError\n",
      "            ExecError\n",
      "            ReadError\n",
      "            URLError\n",
      "                HTTPError\n",
      "                ContentTooShortError\n",
      "            BadGzipFile\n",
      "        EOFError\n",
      "            IncompleteReadError\n",
      "        RuntimeError\n",
      "            RecursionError\n",
      "            NotImplementedError\n",
      "                ZMQVersionError\n",
      "                StdinNotImplementedError\n",
      "            _DeadlockError\n",
      "            BrokenBarrierError\n",
      "            BrokenExecutor\n",
      "                BrokenThreadPool\n",
      "            SendfileNotAvailableError\n",
      "            ExtractionError\n",
      "            VariableError\n",
      "        NameError\n",
      "            UnboundLocalError\n",
      "        AttributeError\n",
      "            FrozenInstanceError\n",
      "        SyntaxError\n",
      "            IndentationError\n",
      "                TabError\n",
      "        LookupError\n",
      "            IndexError\n",
      "            KeyError\n",
      "                NoSuchKernel\n",
      "                UnknownBackend\n",
      "            CodecRegistryError\n",
      "        ValueError\n",
      "            UnicodeError\n",
      "                UnicodeEncodeError\n",
      "                UnicodeDecodeError\n",
      "                UnicodeTranslateError\n",
      "            UnsupportedOperation\n",
      "            JSONDecodeError\n",
      "            SSLCertVerificationError\n",
      "            Error\n",
      "            UnsupportedDigestmodError\n",
      "            IllegalMonthError\n",
      "            IllegalWeekdayError\n",
      "            ParserError\n",
      "            ClassNotFound\n",
      "            ClipboardEmpty\n",
      "            MessageDefect\n",
      "                NoBoundaryInMultipartDefect\n",
      "                StartBoundaryNotFoundDefect\n",
      "                CloseBoundaryNotFoundDefect\n",
      "                FirstHeaderLineIsContinuationDefect\n",
      "                MisplacedEnvelopeHeaderDefect\n",
      "                MissingHeaderBodySeparatorDefect\n",
      "                MultipartInvariantViolationDefect\n",
      "                InvalidMultipartContentTransferEncodingDefect\n",
      "                UndecodableBytesDefect\n",
      "                InvalidBase64PaddingDefect\n",
      "                InvalidBase64CharactersDefect\n",
      "                InvalidBase64LengthDefect\n",
      "                HeaderDefect\n",
      "                    InvalidHeaderDefect\n",
      "                    HeaderMissingRequiredValue\n",
      "                    NonPrintableDefect\n",
      "                    ObsoleteHeaderDefect\n",
      "                    NonASCIILocalPartDefect\n",
      "                    InvalidDateDefect\n",
      "            MacroToEdit\n",
      "            InvalidFileException\n",
      "            UnequalIterablesError\n",
      "            InvalidVersion\n",
      "            _InvalidELFFileHeader\n",
      "            InvalidWheelFilename\n",
      "            InvalidSdistFilename\n",
      "            InvalidSpecifier\n",
      "            InvalidMarker\n",
      "            UndefinedComparison\n",
      "            UndefinedEnvironmentName\n",
      "            InvalidRequirement\n",
      "                RequirementParseError\n",
      "            InvalidVersion\n",
      "        AssertionError\n",
      "        ArithmeticError\n",
      "            FloatingPointError\n",
      "            OverflowError\n",
      "            ZeroDivisionError\n",
      "                DivisionByZero\n",
      "                DivisionUndefined\n",
      "            DecimalException\n",
      "                Clamped\n",
      "                Rounded\n",
      "                    Underflow\n",
      "                    Overflow\n",
      "                Inexact\n",
      "                    Underflow\n",
      "                    Overflow\n",
      "                Subnormal\n",
      "                    Underflow\n",
      "                DivisionByZero\n",
      "                FloatOperation\n",
      "                InvalidOperation\n",
      "                    ConversionSyntax\n",
      "                    DivisionImpossible\n",
      "                    DivisionUndefined\n",
      "                    InvalidContext\n",
      "        SystemError\n",
      "            CodecRegistryError\n",
      "        ReferenceError\n",
      "        MemoryError\n",
      "        BufferError\n",
      "        Warning\n",
      "            UserWarning\n",
      "                GetPassWarning\n",
      "                FormatterWarning\n",
      "            EncodingWarning\n",
      "            DeprecationWarning\n",
      "                ProvisionalWarning\n",
      "            PendingDeprecationWarning\n",
      "            SyntaxWarning\n",
      "            RuntimeWarning\n",
      "                ProactorSelectorThreadWarning\n",
      "                UnknownTimezoneWarning\n",
      "                PEP440Warning\n",
      "            FutureWarning\n",
      "                ProvisionalCompleterWarning\n",
      "            ImportWarning\n",
      "            UnicodeWarning\n",
      "            BytesWarning\n",
      "            ResourceWarning\n",
      "            DeprecatedTzFormatWarning\n",
      "            PkgResourcesDeprecationWarning\n",
      "        _OptionError\n",
      "        _Error\n",
      "        error\n",
      "        Verbose\n",
      "        Error\n",
      "        SubprocessError\n",
      "            CalledProcessError\n",
      "            TimeoutExpired\n",
      "        TokenError\n",
      "        StopTokenizing\n",
      "        ClassFoundException\n",
      "        EndOfBlock\n",
      "        TraitError\n",
      "        Error\n",
      "        Error\n",
      "            CancelledError\n",
      "            TimeoutError\n",
      "            InvalidStateError\n",
      "        _GiveupOnSendfile\n",
      "        error\n",
      "        Incomplete\n",
      "        TimeoutError\n",
      "        InvalidStateError\n",
      "        LimitOverrunError\n",
      "        QueueEmpty\n",
      "        QueueFull\n",
      "        Empty\n",
      "        Full\n",
      "        ArgumentError\n",
      "        ZMQBaseError\n",
      "            ZMQError\n",
      "                ContextTerminated\n",
      "                Again\n",
      "                InterruptedSystemCall\n",
      "            ZMQBindError\n",
      "            NotDone\n",
      "        PickleError\n",
      "            PicklingError\n",
      "            UnpicklingError\n",
      "        _Stop\n",
      "        ArgumentError\n",
      "        ArgumentTypeError\n",
      "        ConfigError\n",
      "            ConfigLoaderError\n",
      "                ArgumentError\n",
      "            ConfigFileNotFound\n",
      "        ConfigurableError\n",
      "            MultipleInstanceError\n",
      "        ApplicationError\n",
      "        error\n",
      "        TimeoutError\n",
      "        error\n",
      "        ReturnValueIgnoredError\n",
      "        KeyReuseError\n",
      "        UnknownKeyError\n",
      "        LeakedCallbackError\n",
      "        BadYieldError\n",
      "        ReturnValueIgnoredError\n",
      "        Return\n",
      "        InvalidPortNumber\n",
      "        error\n",
      "        LZMAError\n",
      "        RegistryError\n",
      "        _GiveupOnFastCopy\n",
      "        Error\n",
      "            NoSectionError\n",
      "            DuplicateSectionError\n",
      "            DuplicateOptionError\n",
      "            NoOptionError\n",
      "            InterpolationError\n",
      "                InterpolationMissingOptionError\n",
      "                InterpolationSyntaxError\n",
      "                InterpolationDepthError\n",
      "            ParsingError\n",
      "                MissingSectionHeaderError\n",
      "        NoIPAddresses\n",
      "        BadZipFile\n",
      "        LargeZipFile\n",
      "        BadEntryPoint\n",
      "        NoSuchEntryPoint\n",
      "        DuplicateKernelError\n",
      "        ErrorDuringImport\n",
      "        NotOneValueFound\n",
      "        CannotEval\n",
      "        OptionError\n",
      "        BdbQuit\n",
      "        Restart\n",
      "        ExceptionPexpect\n",
      "            EOF\n",
      "            TIMEOUT\n",
      "        PtyProcessError\n",
      "        FindCmdError\n",
      "        HomeDirError\n",
      "        ProfileDirError\n",
      "        IPythonCoreError\n",
      "            TryNext\n",
      "            UsageError\n",
      "            StdinNotImplementedError\n",
      "        InputRejected\n",
      "        GetoptError\n",
      "        ErrorToken\n",
      "        PrefilterError\n",
      "        AliasError\n",
      "            InvalidAliasError\n",
      "        Error\n",
      "            InterfaceError\n",
      "            DatabaseError\n",
      "                InternalError\n",
      "                OperationalError\n",
      "                ProgrammingError\n",
      "                IntegrityError\n",
      "                DataError\n",
      "                NotSupportedError\n",
      "        Warning\n",
      "        SpaceInInput\n",
      "        DOMException\n",
      "            IndexSizeErr\n",
      "            DomstringSizeErr\n",
      "            HierarchyRequestErr\n",
      "            WrongDocumentErr\n",
      "            InvalidCharacterErr\n",
      "            NoDataAllowedErr\n",
      "            NoModificationAllowedErr\n",
      "            NotFoundErr\n",
      "            NotSupportedErr\n",
      "            InuseAttributeErr\n",
      "            InvalidStateErr\n",
      "            SyntaxErr\n",
      "            InvalidModificationErr\n",
      "            NamespaceErr\n",
      "            InvalidAccessErr\n",
      "            ValidationErr\n",
      "        ValidationError\n",
      "        EditReadOnlyBuffer\n",
      "        _Retry\n",
      "        InvalidLayoutError\n",
      "        HeightIsUnknownError\n",
      "        ParserSyntaxError\n",
      "        InternalParseError\n",
      "        _PositionUpdatingFinished\n",
      "        SimpleGetItemNotFound\n",
      "        UncaughtAttributeError\n",
      "        HasNoContext\n",
      "        ParamIssue\n",
      "        _JediError\n",
      "            InternalError\n",
      "            WrongVersion\n",
      "            RefactoringError\n",
      "        OnErrorLeaf\n",
      "        InvalidPythonEnvironment\n",
      "        MessageError\n",
      "            MessageParseError\n",
      "                HeaderParseError\n",
      "                BoundaryError\n",
      "            MultipartConversionError\n",
      "            CharsetError\n",
      "        Error\n",
      "        HTTPException\n",
      "            NotConnected\n",
      "            InvalidURL\n",
      "            UnknownProtocol\n",
      "            UnknownTransferEncoding\n",
      "            UnimplementedFileMode\n",
      "            IncompleteRead\n",
      "            ImproperConnectionState\n",
      "                CannotSendRequest\n",
      "                CannotSendHeader\n",
      "                ResponseNotReady\n",
      "            BadStatusLine\n",
      "                RemoteDisconnected\n",
      "            LineTooLong\n",
      "        InteractivelyDefined\n",
      "        KillEmbedded\n",
      "        Error\n",
      "            NoSuchProcess\n",
      "                ZombieProcess\n",
      "            AccessDenied\n",
      "            TimeoutExpired\n",
      "        _Ipv6UnsupportedError\n",
      "        QueueEmpty\n",
      "        QueueFull\n",
      "        DebuggerInitializationError\n",
      "        ExpatError\n",
      "        Error\n",
      "            ProtocolError\n",
      "            ResponseError\n",
      "            Fault\n",
      "        ParseBaseException\n",
      "            ParseException\n",
      "            ParseFatalException\n",
      "                ParseSyntaxException\n",
      "        RecursiveGrammarException\n",
      "        ResolutionError\n",
      "            VersionConflict\n",
      "                ContextualVersionConflict\n",
      "            DistributionNotFound\n",
      "            UnknownExtra\n",
      "        _Error\n",
      "        UnableToResolveVariableException\n",
      "        InvalidTypeInArgsException\n",
      "    GeneratorExit\n",
      "    SystemExit\n",
      "    KeyboardInterrupt\n",
      "    CancelledError\n",
      "    AbortThread\n"
     ]
    }
   ],
   "source": [
    "import inspect\n",
    "import builtins\n",
    "\n",
    "def print_exception_hierarchy(exception_class, indent=0):\n",
    "    print(' ' * indent + exception_class.__name__)\n",
    "    subclasses = exception_class.__subclasses__()\n",
    "    for subclass in subclasses:\n",
    "        print_exception_hierarchy(subclass, indent + 4)\n",
    "\n",
    "# Get the base exception class\n",
    "base_exception = BaseException\n",
    "\n",
    "# Print the exception hierarchy\n",
    "print_exception_hierarchy(base_exception)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d19dfd5-ed45-4f2c-9033-e6e803d61b3f",
   "metadata": {},
   "source": [
    "Q3. What errors are defined in the ArithmeticError class? Explain any two with an example."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a95870f-47ce-4323-8cd7-ff98527dbe00",
   "metadata": {},
   "source": [
    "#### The ArithmeticError class in Python is the base class for all arithmetic-related exceptions. It encompasses errors that occur during arithmetic operations, such as division by zero, invalid operations, or overflow/underflow conditions, value error. Here are two commonly encountered errors defined in the ArithmeticError class, along with examples:\n",
    "\n",
    "#### 1) ZeroDivisionError: This error occurs when a division or modulo operation is performed with zero as the divisor."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "cc1c63a8-8f73-498d-9edc-eef5dc3bd27d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "error: division by zero\n"
     ]
    }
   ],
   "source": [
    "try:\n",
    "    result=10/0\n",
    "except ZeroDivisionError as e:\n",
    "    print(\"error:\",e)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b23a101a-166c-4a61-ba8d-c54bc4645731",
   "metadata": {},
   "source": [
    "#### 2) ValueError: Although ValueError is not directly derived from ArithmeticError, it is often related to arithmetic operations. This error occurs when a function receives an argument of the correct type but an inappropriate value."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "bf5168fe-837e-487d-93dc-2d7261d85e8c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "error: invalid literal for int() with base 10: 'abc'\n"
     ]
    }
   ],
   "source": [
    "try:\n",
    "    result=int(\"abc\")\n",
    "except ValueError as e:\n",
    "    print(\"error:\",e)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce26bd90-1876-46f9-add6-5730b20c7300",
   "metadata": {},
   "source": [
    "Q4. Why LookupError class is used? Explain with an example KeyError and IndexError."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5269902c-2c82-41c3-81b6-7b5c71ed78de",
   "metadata": {},
   "source": [
    "#### The LookupError class in Python is the base class for exceptions that occur when a lookup or indexing operation fails. It is a subclass of the Exception class and provides a common base for exceptions related to lookup operations. The two commonly encountered exceptions that derive from LookupError are KeyError and IndexError.\n",
    "\n",
    "#### 1) KeyError: This exception is raised when a dictionary key is not found."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "edcdc49a-ee80-481e-a13e-1985029c47c7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "error 'city'\n"
     ]
    }
   ],
   "source": [
    "try:\n",
    "    dict={\"name\":\"praju\",\"age\":23}\n",
    "    dict[\"city\"]\n",
    "except KeyError as e:\n",
    "    print(\"error:\",e)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "466fdb83-85b5-4de0-bd11-409505091694",
   "metadata": {},
   "source": [
    "#### 2) IndexError: This exception is raised when an index is out of range in a sequence (e.g., list, tuple, string)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "5fde83a0-03fb-4495-bb8d-7d3e4bec4bea",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "error: list index out of range\n"
     ]
    }
   ],
   "source": [
    "try:\n",
    "    l=[1,2,5,8]\n",
    "    l[6]\n",
    "except IndexError as e:\n",
    "    print(\"error:\",e)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "efd6bfb3-228b-456b-84a7-1eafb1e4dba4",
   "metadata": {},
   "source": [
    "Q5. Explain ImportError. What is ModuleNotFoundError?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63e730b6-bc19-41dc-a35a-ab9a59711e86",
   "metadata": {},
   "source": [
    "#### In Python, ImportError is an exception that is raised when an import statement fails to import a module or a specific name from a module. It is a subclass of the Exception class and is used to handle import-related errors.\n",
    "#### The ImportError exception can occur due to various reasons, such as:\n",
    "\n",
    "1) The module or name being imported does not exist.\n",
    "2) There is an issue with the module's code or structure.\n",
    "3) The module or its dependencies are not installed or accessible.\n",
    "4)  is a circular import or dependency loop between modules.\n",
    "\n",
    "#### e.g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "ff76ef03-5095-4e0e-ba3c-4ab512f3f8ca",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "error: No module named 'non_existent_module'\n"
     ]
    }
   ],
   "source": [
    "try:\n",
    "    import non_existent_module  \n",
    "except ImportError as e:\n",
    "    print(\"error:\", e)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49ec2673-c828-4cea-8a83-464f821044a7",
   "metadata": {},
   "source": [
    "#### However, starting from Python 3.6, there is a more specific subclass of ImportError called ModuleNotFoundError. It is raised when a module or package is not found during the import process. It provides a clearer and more specific error message to indicate that the requested module or package could not be located."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "ab3f8d3b-59ce-4931-a97f-97df61f4941e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "error: No module named 'non_existent_module'\n"
     ]
    }
   ],
   "source": [
    "try:\n",
    "    import non_existent_module  \n",
    "except ModuleNotFoundError as e:\n",
    "    print(\"error:\", e)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46a4d16f-5685-44df-9617-a4fded168bf3",
   "metadata": {},
   "source": [
    "Q6. List down some best practices for exception handling in python."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40a5a929-9214-4279-8824-c58148956fc3",
   "metadata": {},
   "source": [
    "#### The  some best practices for exception handling in Python:\n",
    "\n",
    "#### 1) Be Specific with Exception Types: Catch exceptions at the appropriate level of granularity by specifying the specific exception types to catch. This allows to handle different exceptions differently and avoids catching more exceptions than intended.\n",
    "\n",
    "#### 2) Use Multiple Except Blocks: When handling multiple exceptions, use separate except blocks for each exception type. This makes the code more readable and allows for different handling logic for different exceptions.\n",
    "\n",
    "#### 3) Avoid Bare except: Avoid using bare except statements (i.e., except: without specifying the exception type). Catching all exceptions can make it difficult to identify and debug specific issues. Instead, be explicit about the exceptions we want to handle.\n",
    "\n",
    "#### 4) Use finally for Cleanup: Use the finally block to ensure that critical cleanup code is executed, regardless of whether an exception occurred or not. This is useful for releasing resources or closing files, database connections, or network connections.\n",
    "\n",
    "#### 5) Handle Exceptions Locally: Ideally, handle exceptions as close to the source of the exception as possible. This allows to provide more specific and meaningful error messages and enables better error isolation and debugging."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
